### ISSUE TITLE
Requesting for training scores along with test scores after training models, tuning models or comparing models etc

### ISSUE THREAD

**Issue by levelalphaone** [pycaret/pycaret#1909]
1. Currently, after running the functions create_model, compare_model function, etc., do not return train scores, only test scores across the splits or the mean of the test scores are displayed. 
2. What if someone wants to validate the scores coming in the training sets, or how the scores are varying across the training sets? The comparison between training scores and testing scores helps us understand how the model is performing in terms of fitting, for example, overfitting, underfitting, or covariate shift. This is very crucial in the case of Time aware cross-validation.  
3. I have introduced a parameter namely return_train_score, which is a boolean variable and helps the user to specify if they want the training score to return or not. So, the choice is optional and the training score is merged with the test data frame if the parameter is set as true. I have also updated the display formats accordingly and the scores are shown as Train_Score, Test_Score, etc. 

![image](https://user-images.githubusercontent.com/84379260/144364146-2b766392-568f-4e7e-9a45-4cab0301eb61.png)    

4. After the modification, the function has the outputs in the following manner, the number of scores can easily be reduced with Pycaret's own remove_metric function.
    
![image](https://user-images.githubusercontent.com/84379260/144365655-f1b2bef3-961b-4255-8284-3ac0f3628be8.png)
    
![image](https://user-images.githubusercontent.com/84379260/144365734-ccdcf104-003e-435a-9dfd-92d3e23ebe4c.png)
    
5. If you find this useful, I can dedicate my time to work on this issue and any suggestions on this issue are most welcome.

**Comment by ngupta23** [comment]
This would be useful functionality. But we need to think about the interface since it will apply to `create_model` as well.
May I also recommend adding it in "tall" format with a column for identifying the split (train or test)? That way it is easier to filter if we `pull()` the data frame metrics for further analysis and the user won't have to mess with extracting the splits from the column names.

So columns would be: **Model, Split, Accuracy, AUC, Recall, Precision, F1, ...**

  


@moezali1 @Yard1 - your thoughts?

**Comment by levelalphaone** [comment]
Hi. I could not understand your point. Can you explain via an example maybe? With some images as well? @ngupta23 it would be really helpful.

**Comment by moezali1** [comment]
@levelalphaone This is very useful. I would love to have this go in the next release. (in 1-2 weeks). Can you please clarify what is Train_ score is? Is it mean a cross-validated score, if yes, then it's not really performance of the train set, it is a cross-validated (out of sample) measurement.

**Comment by levelalphaone** [comment]
Hi @moezali1. You are right. The Train_Score does not signify the performance of the model on the entire training set but the training samples of the cross-validation splits. Here is the example - 

![image](https://user-images.githubusercontent.com/84379260/144810488-e7148236-825b-4bdc-8ff4-258ab44861d7.png)

**Comment by ngupta23** [comment]
@moezali1 @levelalphaone @Yard1 

I think we should really consider the format of the output. The words "Train" and "Test" have no place in the columns. 
We will have to change the sort argument from `accuracy` to "test_accuracy"

![image](https://user-images.githubusercontent.com/33585645/150988518-ed6dccac-7f45-4775-b9c6-5b2ade4b1ac7.png)

My strong opinion is that these should be in the tall format and the splits should be row identifiers such as follows. The sort argument should remain = "accuracy" and should look at test only. Is there a reason why someone would pick the model based on train scores?

| Index | Split | Accuracy | AUC  | ... |
|-------|-------|----------|------|-----|
| 0     | Train | 0.9      | 0.85 |     |
| 0     | Test  | 0.85     | 0.86 |     |
| 1     | Train | 0.6      | 0.9  |     |
| 1     | Test  | 0.65     | 0.92 |     |
| 2     | Train | 0.8      | 0.9  |     |
| 2     | Test  | 0.82     | 0.87 |     |
| ...   |       |          |      |     |

This would be easier to manage.

**Comment by levelalphaone** [comment]
@ngupta23 We can sure do that, but this only gives the choice to the user how he wants to sort, optimize his model. If we need to sort or optimize only on the metric names only, can you please tell me the name of the functions, or if in all the functions, then can you please respond accordingly? 
@moezali1 Please comment on this as well.

**Comment by ngupta23** [comment]
@levelalphaone There may be more nuances to this now that I think of it. For example, you are not really showing the test scores here - what you are showing is the train and validation part of the Cross validation folds. 

In addition, we also can have the scores on the entire train dataset (trained model after cv) as well as the test score (not used in training at all). 

Let me think about this some more, discuss with the team and get back.

**Comment by levelalphaone** [comment]
@ngupta23 you are absolutely right. I have mentioned in this comment thread, that the test scores here are nothing but the validation scores. Please let me know what other changes I have to make for this PR. Thanks again!

**Comment by ngupta23** [comment]
Some more thoughts for discussion with team

| Index |     Split    | Accuracy |  AUC | ... |
|:-----:|:------------:|:--------:|:----:|:---:|
| 0     | CV-Train     | 0.9      | 0.85 |     |
| 0     | CV-Val       | 0.85     | 0.86 |     |
| 1     | CV-Train     | 0.6      | 0.9  |     |
| 1     | CV-Val       | 0.65     | 0.92 |     |
| 2     | CV-Train     | 0.8      | 0.9  |     |
| 2     | CV-Val       | 0.82     | 0.87 |     |
| ...   |              |          |      |     |
| Mean  | CV-Train     | 0.8      | 0.9  |     |
| Std   | CV-Train     | 0.1      | 0.05 |     |
| Mean  | CV-Val       | 0.78     | 0.89 |     |
| Std   | CV-Val       | 0.12     | 0.05 |     |
|       | Train        | 0.82     | 0.87 |     |
|       | Test/Holdout | 0.84     | 0.88 |     |

**Comment by ngupta23** [comment]
@levelalphaone We discussed this in an internal meeting today and this is what we are recommending as the final structure. This structure will enable easy filtering and plotting of the results to diagnose under/overfitting in the models as shown below. Can you update your code to implement this?

Note that the last line with `Split = Train` is the score for the final model that is returned from `create_model` (and other functions that return a single model) after CV is done and the model is trained on the entire train dataset. Let me know if you have any questions.

### Example Code

https://gist.github.com/ngupta23/86cc7cc3221693e71a45b6a4431be104

### Recommended Structure

![image](https://user-images.githubusercontent.com/33585645/153719607-a07c09ad-1028-4a90-b0a8-1d0432b5868a.png)

### Diagnosing Under/Overfitting

![image](https://user-images.githubusercontent.com/33585645/153719463-bc999b54-9f87-4cdb-9078-ee1be9f900e4.png)

**Comment by levelalphaone** [comment]
@ngupta23 Sure! This looks really clean and has a much better implementation. I will try to close this within two weeks, and I will post my doubts here if I come into any. Thanks for the information and support.

**Comment by ngupta23** [comment]
No problem. Just let us know if you face any issues. Thanks for your work on this PR.

**Comment by moezali1** [comment]
@levelalphaone @ngupta23 I am thinking we can do 2.3.7 release this week. I would love this to be in 2.3.7. Do you think it is possible to get PR submitted and approved by Thu/Fri?

Thanks.

**Comment by levelalphaone** [comment]
Sorry @moezali1 for replying this late. I will work on this and create appropriate PR as soon as possible.

**Comment by levelalphaone** [comment]
Should we give the option for sorting or optimizing for train metrics or only for test metrics, while doing compare_model, automl, or similar functions? Can you help me with this @ngupta23?

**Comment by ngupta23** [comment]
> Should we give the option for sorting or optimizing for train metrics or only for test metrics, while doing compare_model, automl, or similar functions? 

It should never be based on the test metric as that will be leakage. It should always be based on the CV-Val score. Let's just keep it as such for now. Hope that helps.

**Comment by levelalphaone** [comment]
@ngupta23 sorry for the miscommunication. By test, I wanted to mention CV-Val only. Thanks for the clarification.

**Comment by levelalphaone** [comment]
@ngupta23 Can you please review the code to the new pull request I just put in? Modified and mentioned some comments about two test files as well. And please let me know if everything is okay. Thank you.
